marshal/reflect_cache.go

```diff
@@ -0,0 +1,37 @@
+package marshal
+
+import "reflect"
+
+var typeCache = map[reflect.Type]*[]cachedType{}
+
+var nameCache = map[reflect.Type]string{}
+
+type cachedType struct {
+	reflect.StructField
+	JsonTag string
+}
+
+func getTypeFields(structType reflect.Type) *[]cachedType {
+	if res, ok := typeCache[structType]; ok {
```

Comment:
No need to use a pointer type for the map value or return type. A slice copy is always by reference.